### ISSUE TITLE
Doesn't work as expected when programmatically calling babel

### ISSUE THREAD

**Issue by petejodo** [michalkvasnicak/babel-plugin-css-modules-transform#45]
I'm making a custom build process that handles calling babel-core's `transform` itself for each file, each time with a custom `extractCss` property. This plugin looks like it's setting it's options only once instead of each time `babel.transform` is called so effectively the options passed to that call are actually irrelevant except for the 1st call.

Separately, there should be a way to print the css modules errors instead of just reporting nothing. It makes it difficult to debug css issues. I'd be willing to work on the PR but figured there could be a discussion first

**Comment by PeteJodo** [comment]
also the `post` function runs after each call to `babel.transform` so for each consecutive call, it will write all .css files processed up to that point

**Comment by PeteJodo** [comment]
For anyone else with this issue, a temporary solution is to reset the plugin before each call so something like:

```javascript
const pathToPlugin = path.resolve(
    process.cwd(),
    'node_modules/babel-plugin-css-modules-transform/build/index.js'
);
if (!!require.cache[pathToPlugin]) {
    delete require.cache[pathToPlugin];
}
```

**Comment by michalkvasnicak** [comment]
@PeteJodo Thank you! It is caused because:

1. this plugin can export css file so it has to somehow keep a track of compiled files
2. this plugin relies on `css-modules-require-hook` so basically options are passed to this hook.

This plugin was not intended to be used outside of babel-cli.

> also the post function runs after each call to babel.transform so for each consecutive call, it will write all .css files processed up to that point which ends up being a lot of duplicate work

Yes because when you start babel-cli in normal or watch mode you don't have any way how to detect if compilation ended so I need to write files on every post compilation. At least I haven't found a way how to do it properly. I'm open to any ideas how to do it properly.

> Separately, there should be a way to print the css modules errors instead of just reporting nothing. It makes it difficult to debug css issues. I'd be willing to work on the PR but figured there could be a discussion first

This is strange because I'm just requiring css files, so if `css-modules-require-hook` does not throw an error I don't have any way to report it. But maybe I have a bug here and I suppress errors.

**Comment by PeteJodo** [comment]
Hey thanks for answering.

> Yes because when you start babel-cli in normal or watch mode you don't have any way how to detect if compilation ended so I need to write files on every post compilation. At least I haven't found a way how to do it properly. I'm open to any ideas how to do it properly.

Yeah I see the issue here. I'd need to do research myself about the differences between `babel.transform` and `babel-cli` so I can't really comment on that either.

> 1. this plugin can export css file so it has to somehow keep a track of compiled files
> 2. this plugin relies on `css-modules-require-hook` so basically options are passed to this hook.

It's not so much the config being passed to the require-hook that's the issue, it's the `extractCss` property which looks like the plugin's config and not a css-modules config. `extractCss` could be set in `pre` but ergonomically, it would be weird that only part of the passed-in config would update on each call to `babel.transform`. Personally I don't think/expect a user to change the css-modules portion of the config (or at least I don't think it's a good idea anyway).

> This is strange because I'm just requiring css files, so if `css-modules-require-hook` does not throw an error I don't have any way to report it. But maybe I have a bug here and I suppress errors.

So the way I've been temporarily doing this so that I can see any errors (i.e. an incorrect scss `@import` path) is just editing the built `build/index.js` in my node_modules, adding `console.log(e)` [in this try/catch](https://github.com/michalkvasnicak/babel-plugin-css-modules-transform/blob/master/src/index.js#L39-L41)

Thanks again!

**Comment by PeteJodo** [comment]
@michalkvasnicak bumping specifically for this point...
> > This is strange because I'm just requiring css files, so if `css-modules-require-hook` does not throw an error I don't have any way to report it. But maybe I have a bug here and I suppress errors.
>
> So the way I've been temporarily doing this so that I can see any errors (i.e. an incorrect scss `@import` path) is just editing the built `build/index.js` in my node_modules, adding `console.log(e)` [in this try/catch](https://github.com/michalkvasnicak/babel-plugin-css-modules-transform/blob/master/src/index.js#L39-L41). **I figure something could error out here but I don't know the behavior when using just through babel-cli**

Any errors caused by either `generateScopedName`, `preprocessCss`, and any other extension that is passed off to `css-modules-require-hook` gets caught in [that try/catch](https://github.com/michalkvasnicak/babel-plugin-css-modules-transform/blob/master/src/index.js#L39-L41).

**Comment by PeteJodo** [comment]
Also about your response...

> Yes because when you start babel-cli in normal or watch mode you don't have any way how to detect if compilation ended so I need to write files on every post compilation. At least I haven't found a way how to do it properly. I'm open to any ideas how to do it properly.

why doesn't

**Comment by PeteJodo** [comment]
sorry, didn't mean to close.... pressed wrong button >.<

**Comment by michalkvasnicak** [comment]
Ok so basically we need to detect if the error is from `ignoring` file or is normal error. I will take a look but I have free time only on weekend.

**Comment by PeteJodo** [comment]
Looking at `css-modules-require-hook`, they have this [exception checker](https://github.com/css-modules/css-modules-require-hook/blob/master/lib/index.js#L151-L161) which gets passed into their `attachHook` method which gets used [here](https://github.com/css-modules/css-modules-require-hook/blob/master/lib/attachHook.js#L12).

It looks like by ignoring it, they assume they can just call the existing require extension but I feel like there should be a check there to see if the require extension even exists. I think that's why there's errors thrown for ignored files...because Node doesn't know how to import it.

This is just from glancing at their code. My knowledge on `require.extensions` isn't extensive so I could be mistaken

**Comment by PeteJodo** [comment]
I confirmed via your test suite that this is exactly what is happening. I'm going to open a PR on `css-modules-require-hook` and then once that's in, you can just throw all errors

**Comment by PeteJodo** [comment]
actually the error may be intended... I started a discussion [here](https://github.com/css-modules/css-modules-require-hook/issues/43)

**Comment by michalkvasnicak** [comment]
@PeteJodo maybe if we catch only the error with specific message and rethrow all other errors? But this solution is error prone by itself.

**Comment by PeteJodo** [comment]
yeah I don't like that solution. Currently, the only thing you could do really is...
```js
if (e.message === 'existingHook is not a function') { /* ... */ }
```
I'm going to propose having the require hook throw its own error so it's actually catch-able. That's technically a breaking change for them though (because someone else could be doing the same thing in the above code)

**Comment by PeteJodo** [comment]
I just opened a new issue since I didn't get a response in the closed issue, https://github.com/css-modules/css-modules-require-hook/issues/92
